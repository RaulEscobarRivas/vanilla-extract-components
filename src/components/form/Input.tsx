import React, { useCallback, useContext } from "react";
import classNames from "classnames";
import { inputStyle, inputErrorStyle } from "./Input.css";
import { useInput } from "../../hooks/useInput";
import { InputContext } from "./context";
import { InputProps } from "./types";

export function Input(props: InputProps) {
  const {
    className,
    initialValue,
    isInvalid,
    onChange,
    onBlur,
    onFocus,
    ...otherProps
  } = props;
  const { setIsFocused } = useContext(InputContext);
  const inputClasses = classNames(inputStyle, className, {
    [inputErrorStyle]: isInvalid,
  });

  const handleBlur = useCallback(
    (event) => {
      if (onBlur) {
        onBlur(event);
      }

      // Here we inform the InputGroup that we have lost focus.
      if (setIsFocused) {
        setIsFocused(false);
      }
    },
    [onBlur]
  );

  const handleFocus = useCallback(
    (event) => {
      if (onFocus) {
        onFocus(event);
      }

      // Here we inform the InputGroup that we have gained focus.
      if (setIsFocused) {
        setIsFocused(true);
      }
    },
    [onFocus]
  );

  // Here we setup a controlled value to manage the input value. It will be
  // kept in sync with the field value or a control prop.
  const [value, setValue] = useInput({
    value: props.value || initialValue,
    onChange,
  });

  return (
    <input
      {...otherProps}
      className={inputClasses}
      onBlur={handleBlur}
      onChange={setValue}
      onFocus={handleFocus}
      value={value}
    />
  );
}

Input.defaultProps = {
  type: "text",
};
