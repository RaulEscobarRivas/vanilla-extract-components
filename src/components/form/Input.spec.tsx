import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { Input } from './Input';

describe('Input', () => {
  it('should render default style', () => {
    const { container } = render(
      <Input />
    );

    expect(container).toMatchSnapshot();
  });

  it('should render error style', () => {
    const { container } = render(
      <Input isInvalid />
    );

    expect(container).toMatchSnapshot();
  });

  it('should render with control value', () => {
    const { container } = render(
      <Input value="test" />
    );

    expect(container).toMatchSnapshot();
  });

  it('should append className', () => {
    const { container } = render(
      <Input className="myClass" />
    );

    expect(container).toMatchSnapshot();
  });

  it('should call onChange', () => {
    const callback = jest.fn();

    render(
      <Input data-testid="field" onChange={callback} />
    );

    // Simulate the change event.
    fireEvent.change(screen.getByTestId("field"), { target: { value: 'test' } })

    expect(callback).toHaveBeenCalledTimes(1);
  });
});