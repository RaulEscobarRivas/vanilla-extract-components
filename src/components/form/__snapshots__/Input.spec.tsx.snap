// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should append className 1`] = `
<div>
  <input
    class="Input_inputStyle__lwe5cj0 myClass"
    type="text"
    value=""
  />
</div>
`;

exports[`Input should render default style 1`] = `
<div>
  <input
    class="Input_inputStyle__lwe5cj0"
    type="text"
    value=""
  />
</div>
`;

exports[`Input should render error style 1`] = `
<div>
  <input
    class="Input_inputStyle__lwe5cj0 Input_inputErrorStyle__lwe5cj1"
    type="text"
    value=""
  />
</div>
`;

exports[`Input should render with control value 1`] = `
<div>
  <input
    class="Input_inputStyle__lwe5cj0"
    type="text"
    value="test"
  />
</div>
`;
